### PR TITLE
Added applier to automate testing

### PR DESCRIPTION
**Pull request by garethahealy**
#### What is this PR About?
Added applier to automate testing.

#### How do we test this?
./_test/setup.sh applier pipelinelib-testing garethahealy/pipeline-library auto-testing && ./_test/setup.sh test

cc: @redhat-cop/day-in-the-life

### CODE REVIEW

File: _test/setup.sh [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    fi
+  done
+  return 0
+}
+
+# Process arguments
+case $1 in
+  cluster_up)
+    cluster_up
+    ;;
```
Comment (pabrahamsson) [inline _test/setup.sh]: Same as above

File: _test/setup.sh [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    if [ "$?" -eq 0 ]; then
+      built=true
+      break
+    fi
+    echo "Retrying oc cluster up after failure"
+    oc cluster down
+    sleep 5
+  done
+  echo "OpenShift Cluster Running"
+}
```
Comment (pabrahamsson) [inline _test/setup.sh]: I think this section can be removed as `oc cluster up` is no longer supported in OCP4

File: _test/setup.sh [hunk 3]
```diff
@@ -0,0 +1,122 @@
+#!/bin/bash
+trap "exit 1" TERM
+export TOP_PID=$$
+NAMESPACE="${2:-pipelinelib-testing}"
+TRAVIS_REPO_SLUG="${3:-redhat-cop/pipeline-library}"
+TRAVIS_BRANCH="${4:-master}"
```
Comment (pabrahamsson) [inline _test/setup.sh]: I would suggest removing `TRAVIS_` from these variables in the event we aren't using Travis CI (or for local testing).

File: .applier/group_vars/seed-hosts.yml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    OC_TOKEN_B64: "{{ oc_token | b64encode }}"
+    REGISTRY_URL_B64: "{{ internal_registry_url.split('/')[0] | b64encode }}"
+    DOCKERCFG_B64: "{{ secrets.dockercfg_json | to_json | b64encode }}"
+  tests:
+    files: "{{ lookup('fileglob', '{{ inventory_dir }}/../test/Jenkinsfile-*', wantlist=true) }}"
+
+openshift_cluster_content:
+- object: Environment Setup
+  content:
+    - name: Create Projects
```
Comment (pabrahamsson) [inline .applier/group_vars/seed-hosts.yml]: I know yaml isn't forcing us to use indentation for lists but lets pick one them.

File: .openshift/templates/jenkins-imagestreams.yml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  metadata:
+    labels:
+      role: jenkins-slave
+    name: jenkins-slave-image-mgmt
+  spec:
+    tags:
+      - from:
+          kind: DockerImage
+          name: siamaksade/jenkins-slave-skopeo-centos7
+        name: latest
```
Comment (pabrahamsson) [inline .openshift/templates/jenkins-imagestreams.yml]: Same as above regarding yaml list indentation

File: .applier/group_vars/seed-hosts.yml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        - jenkins-imagestreams
+- object: jenkins-persistent
+  content:
+    - name: jenkins-persistent
+      template: "https://raw.githubusercontent.com/redhat-cop/openshift-templates/{{ templates_repo_ref }}/jenkins/jenkins-persistent-template.yml"
+      params_from_vars: "{{ jenkins.deploy }}"
+      namespace: "{{ namespace }}"
+      tags:
+        - jenkins-persistent
+      # todo: post wait for deployment, Create global pipeline lib, Install sonar plugin
```
Comment (pabrahamsson) [inline .applier/group_vars/seed-hosts.yml]: Is there a requirement for persistent Jenkins or can we use ephemeral?

**Comment by garethahealy** [comment]
@pabrahamsson ; ready for review / merge.

**Comment by pabrahamsson** [comment]
Thank you @garethahealy 
I took it for a spin and it looks great. I ran in to two issues one that seems to happen occasionally and one that's reproducible. The occasional one (happened once) was `verifyDeployment` which never finished (ran for 5 hours), but worked fine on subsequent runs (we might need a timeout).
The reproducible one is `sonarqubeStaticAnalysis` test which fails with 
```
curl --header 'Authorization: Bearer ****' --data-urlencode 'script=import jenkins.model.*
import hudson.model.*
import hudson.plugins.sonar.*
import hudson.tools.*
import hudson.util.Secret
import com.cloudbees.plugins.credentials.impl.*
import com.cloudbees.plugins.credentials.*
import com.cloudbees.plugins.credentials.domains.*
import org.jenkinsci.plugins.plaincredentials.impl.StringCredentialsImpl

def instance = Jenkins.getInstance()

def SONAR_URL = "<let's hide this for now>"
def SONAR_TOKEN = "<let's hide this too>"

def stringCredentials = new StringCredentialsImpl(CredentialsScope.GLOBAL, "sonartoken", null, Secret.fromString("${SONAR_TOKEN}"))

SystemCredentialsProvider systemCredentialsProvider = SystemCredentialsProvider.getInstance()
systemCredentialsProvider.getStore().addCredentials(Domain.global(), stringCredentials)
systemCredentialsProvider.save()

SonarInstallation sonarInstallation = new SonarInstallation("sonar", "${SONAR_URL}", "sonartoken", null, null, null, null, null)

SonarGlobalConfiguration sonarGlobalConfiguration = instance.getDescriptor(SonarGlobalConfiguration.class)
sonarGlobalConfiguration.setInstallations(sonarInstallation)
sonarGlobalConfiguration.migrateCredentials()
sonarGlobalConfiguration.save()

instance.save()' https://<hiding for now>/scriptText
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0groovy.lang.GroovyRuntimeException: Could not find matching constructor for: hudson.plugins.sonar.SonarInstallation(java.lang.String, org.codehaus.groovy.runtime.GStringImpl, java.lang.String, null, null, null, null, null)
	at groovy.lang.MetaClassImpl.invokeConstructor(MetaClassImpl.java:1732)
	at groovy.lang.MetaClassImpl.invokeConstructor(MetaClassImpl.java:1532)
	at org.codehaus.groovy.runtime.callsite.MetaClassConstructorSite.callConstructor(MetaClassConstructorSite.java:49)
...
```
This also fails when I try from the CLI. However, if I change
```
SonarInstallation sonarInstallation = new SonarInstallation("sonar", "${SONAR_URL}", "sonartoken", null, null, null, null, null)
```
to
```
SonarInstallation sonarInstallation = new SonarInstallation("sonar", "${SONAR_URL}", "sonartoken", null, null, null, null)
```
the curl command completes successfully.
According to the [sonar-scanner source](https://github.com/SonarSource/sonar-scanner-jenkins/blob/sonar-2.10/src/main/java/hudson/plugins/sonar/SonarInstallation.java) it should either be 7 or 9 arguments. I wasn't able to get it working with 9 even though Jenkins is reporting `sonar 2.10`. Have you encountered this?

**Comment by garethahealy** [comment]
re: issues 
1. sonar: raised a pr: https://github.com/redhat-cop/pipeline-library/pull/94
2. verify deployment: its a known issue, but i've got something in the works for it: https://github.com/redhat-cop/pipeline-library/issues/95

All my changes have been pushed. ready for final review/merge.

**Pull request merged**